xonsh/xontribs.py

```diff
@@ -76,8 +71,38 @@ def update_context(name, ctx=None):
 @functools.lru_cache()
 def xontrib_metadata():
     """Loads and returns the xontribs.json file."""
-    with open(xontribs_json(), "r") as f:
-        md = json.load(f)
+    impres = None
+    pkg_resources = None
+
+    # NOTE: Reduce all of these alternate implementations when the minimum Python
+    #       is >=3.7
+    try:
+        # Python 3.7
+        import importlib.resources as impres
+    except ImportError:
+        try:
+            # Optional backport for <3.7
+            import importlib_resources as impres
+        except ImportError:
+            try:
+                # Try the slower and clunkier pkg_resources
+                # This is only available if setuptools is part of the environment
+                import pkg_resources
```

Comment:
Is `pkg_resources` really needed to support zipapp? I would like to avoid this package if at all possible.

Reply:
I mean, you could also write the zip-reading and path traversal code yourself. But this already exists in `pkg_resources`, it's stdlib counterpart `importlib.resources`, and the stdlib backport `importlib_resources`. Hence all the variants.

All of these are completely optional dependencies. However, `pkg_resources` is part of `setuptools`, which is a dependency of `pip`, so I expect many real-world environments to include it. (Exceptions being things like some condas and several stock Linux distros.)

In the case where none of the three modules are available, it falls back to using the filename. While this is completely incompatible with any form of zip-based execution, it should mean nothing will break. (And it's easy enough to include extra dependencies when building the zip.)

The [`importlib.resources` docs](https://docs.python.org/3/library/importlib.html#module-importlib.resources) includes some discussion about these modules.

Reply:
There's also a reason why it's my 3rd choice of package to use.

Reply:
Oh, and xonsh implicitly optionally depends on `pkg_resources` because it's the only way I know of to utilize entry points, and xonsh registers entry points with pygments and pytest.

Reply:
Yeah - I know, we have just goine to crazy lengths to avoid importing `pkg_resources` (or importing it in a thread) on startup. My concern is that for folks using xontribs in their xonshrc's, this could really slow down their startup times.  That said, it is worth merging as is for now, and let's keep an eye on it.  Thanks again @astronouth7303!

Reply:
If it's bad, that version can easily be removed.